lib/matplotlib/colors.py

```diff
@@ -62,55 +65,201 @@
+from ._color_data import BASE_COLORS, CSS4_COLORS, XKCD_COLORS
 
-# for back copatibility
-cnames = CSS4_COLORS
 
-COLOR_NAMES = {'xkcd': XKCD_COLORS,
-               'css4': CSS4_COLORS}
+class _ColorMapping(dict):
+    def __init__(self, mapping):
+        super(_ColorMapping, self).__init__(mapping)
+        self._cache = {}
 
+    def __setitem__(self, key, value):
+        super(_ColorMapping, self).__setitem__(key, value)
+        self._cache.clear()
 
-def is_color_like(c):
-    'Return *True* if *c* can be converted to *RGB*'
-
-    # Special-case the N-th color cycle syntax, because its parsing
-    # needs to be deferred.  We may be reading a value from rcParams
-    # here before the color_cycle rcParam has been parsed.
-    if isinstance(c, bytes):
-        match = re.match(b'^C[0-9]$', c)
-        if match is not None:
-            return True
-    elif isinstance(c, six.text_type):
-        match = re.match('^C[0-9]$', c)
-        if match is not None:
-            return True
+    def __delitem__(self, key, value):
+        super(_ColorMapping, self).__delitem__(key, value)
+        self._cache.clear()
 
-    try:
-        colorConverter.to_rgb(c)
+
+_colors_full_map = {}
+# Set by reverse priority order.
+_colors_full_map.update(XKCD_COLORS)
+_colors_full_map.update(CSS4_COLORS)
+_colors_full_map.update(BASE_COLORS)
+_colors_full_map = _ColorMapping(_colors_full_map)
+
+
+def get_named_colors_mapping():
+    """Return the global mapping of names to named colors.
+    """
+    return _colors_full_map
+
+
+def _is_nth_color(c):
+    """Return whether `c` can be interpreted as an item in the color cycle.
+    """
+    return isinstance(c, six.string_types) and re.match(r"\AC[0-9]\Z", c)
+
+
+def is_color_like(c):
+    """Return whether `c` can be interpreted as an RGB(A) color.
+    """
+    # Special-case nth color syntax because it cannot be parsed during
+    # setup.
+    if _is_nth_color(c):
         return True
+    try:
+        to_rgba(c)
     except ValueError:
         return False
+    else:
+        return True
 
 
-def rgb2hex(rgb):
-    'Given an rgb or rgba sequence of 0-1 floats, return the hex string'
-    a = '#%02x%02x%02x' % tuple([int(np.round(val * 255)) for val in rgb[:3]])
-    return a
+def to_rgba(c, alpha=None):
+    """Convert `c` to an RGBA color.
 
+    If `alpha` is not `None`, it forces the alpha value.
+    """
+    # Special-case nth color syntax because it should not be cached.
+    if _is_nth_color(c):
+        from matplotlib import rcParams
+        from matplotlib.rcsetup import cycler
+        prop_cycler = rcParams['axes.prop_cycle']
+        if prop_cycler is None and 'axes.color_cycle' in rcParams:
+            clist = rcParams['axes.color_cycle']
+            prop_cycler = cycler('color', clist)
+        colors = prop_cycler._transpose()['color']
+        c = colors[int(c[1]) % len(colors)]
+    try:
+        rgba = _colors_full_map._cache[c, alpha]
```

Comment:
It is obviously possible to make an alpha-less cache work, but I'm not sure the memory savings are worth it.  On the other hand, that would certainly be a precondition for making the cache public (otherwise it is truly too implementation-specific and should be kept private).